src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/PropertyDescriptor.cs

```diff
@@ -165,10 +170,12 @@ public virtual void AddValueChanged(object component, EventHandler handler)
             ArgumentNullException.ThrowIfNull(component);
             ArgumentNullException.ThrowIfNull(handler);
 
-            _valueChangedHandlers ??= new Dictionary<object, EventHandler?>();
+            if (_valueChangedHandlers is null)
+            {
+                Interlocked.CompareExchange(ref _valueChangedHandlers, new ConcurrentDictionary<object, EventHandler?>(), null);
+            }
 
-            EventHandler? h = _valueChangedHandlers.GetValueOrDefault(component, defaultValue: null);
-            _valueChangedHandlers[component] = (EventHandler?)Delegate.Combine(h, handler);
+            _valueChangedHandlers.AddOrUpdate(component, handler, (k, v) => (EventHandler?)Delegate.Combine(v, handler));
```

Comment:
Can this lock-free AddOrUpdate have bad race conditions with Remove?

Reply:
I think either both Add and Remove need be lock-free as proposed in the other comment, or both need to take the lock.

Reply:
I was just thinking about this as well; I'll revert this since it will collide with the lock in Remove.